### PR TITLE
19: Timesheet creation.

### PR DESCRIPTION
**Pull request by josue2591**
**Steps to test**

1.
 **Before test functionality**
- Check if the fields of timesheet content type were created fine.
- Check if the fields needed by timesheet content type were added to manatime_feature_field_bases.
- Check if the manatime_feature_timesheet was added to settings.php in the right order.
- Created some collaborators users.

2.
 **To test the functionality works**
- Run a variable_set in Execute PHP code of Drupal with a razonable date in timestamp format (It must be a date wich has already passed).
- Run cron.
- Check if the timesheet nodes corresponding to the collaborator users were created.

[Issue link](https://github.com/ManatiCR/manatime/issues/19)

### CODE REVIEW

**Comment by robherba** [comment]
Everything looks great for me!

File: modules/features/manatime_feature_field_bases/manatime_feature_field_bases.info [hunk 1]
```diff
@@ -18,6 +18,8 @@ features[field_base][] = field_duration
 features[field_base][] = field_project
 features[field_base][] = field_status
 features[field_base][] = field_task
-features[field_base][] = field_tasks
```
Comment (kporras07) [inline modules/features/manatime_feature_field_bases/manatime_feature_field_bases.info:21]: I think this line was removed by error.

File: modules/features/manatime_feature_timesheet/manatime_feature_timesheet.install [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+/**
+ * @file
+ * Install some functions for the timesheet feature.
+ */
+
+/**
+ * Implement hook install.
+ */
+function manatime_feature_timesheet_install() {
+  variable_set('manatime_feature_timesheet_execution_time', 1473627600);
```
Comment (kporras07) [inline modules/features/manatime_feature_timesheet/manatime_feature_timesheet.install]: Add a comment explaining what is that time.

File: modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+include_once 'manatime_feature_timesheet.features.inc';
+
+/**
+ * Implements hook_cron().
+ */
+function manatime_feature_timesheet_cron() {
+
+  $time_execution = variable_get('manatime_feature_timesheet_execution_time', 0);
+
+  if (REQUEST_TIME >= $time_execution) {
```
Comment (kporras07) [inline modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module]: Add a comment explaining why are you doing this condition.

File: modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    $next_execution_time = $time_execution + 604800;
```
Comment (kporras07) [inline modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module]: Why 604800? (add a comment)

File: modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    $begining_date = format_date(REQUEST_TIME, 'custom', 'F d, Y');
```
Comment (kporras07) [inline modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module]: I think REQUEST_TIME is not the right beginning date. Request time it's supposed to be sunday, but your beginning date is supposed to be monday.
Btw, typo in variable name.

File: modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    $end_date = format_date($next_execution_time, 'custom', 'F d, Y');
+    $begining_date_field = format_date(REQUEST_TIME, 'custom', 'Y-m-d H:i:s');
```
Comment (kporras07) [inline modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module]: Same here

Comment (kporras07) [inline modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module]: Why this format? (comment)

File: modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    $end_date_field = format_date($next_execution_time, 'custom', 'Y-m-d H:i:s');
+
+    $query = db_select('users_roles', 'ur');
+    $query->join('role', 'r', 'r.rid = ur.rid AND r.name = :name', array(
+      ':name' => 'collaborator',
+    ));
+    $query->fields('ur', array('uid'));
+    $result = $query->execute()->fetchAll();
```
Comment (kporras07) [inline modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module]: Change fetchAll to something that fetch one record at a time.

File: modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    foreach ($result as $user_role) {
+      $uid = $user_role->uid;
+
+      $timesheet = new stdClass();
```
Comment (kporras07) [inline modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module]: After this line, invoke `node_object_prepare($timesheet);`

File: modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      $timesheet->title = 'Timesheet from: ' . $begining_date . ' to: ' . $end_date;
+      $timesheet->type = 'timesheet';
+
+      $timesheet->field_date[LANGUAGE_NONE][0]['value'] = $begining_date_field;
+      $timesheet->field_date[LANGUAGE_NONE][0]['value2'] = $end_date_field;
+      $timesheet->field_user[LANGUAGE_NONE][0]['target_id'] = $uid;
+
+      node_save($timesheet);
+    }
+    Variable_set('manatime_feature_timesheet_execution_time', $next_execution_time);
```
Comment (kporras07) [inline modules/features/manatime_feature_timesheet/manatime_feature_timesheet.module]: this is going to make the world explode: function name is in lower case.

File: settings/settings.php [hunk 10]
```diff
@@ -62,6 +62,7 @@
   'manatime_feature_site_template',
   'manatime_feature_tasks',
   'manatime_feature_time_entries',
+  'manatime_feature_timesheet',
```
Comment (kporras07) [inline settings/settings.php]: For consistency, I'd make the feature name plural. Ping me when you're going to do this to help you.

**Comment by josue2591** [comment]
@kporras07 Done all the comments waiting for review.

File: modules/features/manatime_feature_timesheets/manatime_feature_timesheets.module [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  // Get the variable wich has the execution time hour.
+  $execution_time = variable_get('manatime_feature_timesheets_execution_time', 0);
+  // Conditional to evaluate if it is time to execute this function, the
+  // function must execute itself the sundays at 21:00:00 hours.
+  if (REQUEST_TIME >= $execution_time) {
+    // Get the beginning and end time to use in the new nodes.
+    $beginning = new DateTime('monday 00:00');
+    $end = new DateTime('next sunday 21:00');
+    // Set the next time the function will execute itself in timestamp format.
+    $next_execution_time = $beginning->getTimestamp();
```
Comment (kporras07) [inline modules/features/manatime_feature_timesheets/manatime_feature_timesheets.module]: Actually, I think you should use $end here.

**Comment by kporras07** [comment]
I know for faith this work, however; it would be awesome if you could think of a way to make it more testable. Ex. I set the variable to something after my actual date and executed cron today (tuesday). I got timesheets from next monday (09/19) to sunday (09/18). Maybe calculate end date based on beginning date?

**Comment by kporras07** [comment]
@josue2591 :shipit:

**Pull request merged**